awx/main/scheduler/task_manager_instances.py

```diff
@@ -0,0 +1,204 @@
+                    # TODO: Don't call the Instance.remaining_capacity property because this queries all the running and
+                    # waiting tasks on this instance, which is expensive. Generally there are more tasks than instances
+                    # so it costs less to loop over instances multiple times than query all running and waiting tasks multiple times
+                    # Instead we may need to init with total capacity, and then later when
+                    # we have the tasks update the nodes w/ the right remaining capacity. This more closly matches
+                    # what we used to do tracking capacity remaining on the instance group
+                    PrioritizableNode(remaining_capacity=instance.remaining_capacity, hostname=instance.hostname, jobs_running=instance.jobs_running)
+                )
+                if instance.hostname not in self.instances_partial:
+                    self.instances_partial[instance.hostname] = SimpleNamespace(
+                        obj=instance,
+                        node_type=instance.node_type,
+                        hostname=instance.hostname,
+                        instance_groups=[rampart_group.name],
+                    )
+                else:
+                    self.instances_partial[instance.hostname].instance_groups.append(rampart_group.name)
+            logger.debug(f"Control nodes in {rampart_group.name}: {self.nodes['control_nodes'][rampart_group.name]}")
+            logger.debug(f"Hybrid nodes in {rampart_group.name}: {self.nodes['hybrid_nodes'][rampart_group.name]}")
+            logger.debug(f"Execution nodes {rampart_group.name}: {self.nodes['execution_nodes'][rampart_group.name]}")
+
+    def __getitem__(self, key):
+        return self.instances_partial[key]
+
+    def find_largest_idle_instance(self, instance_group, capacity_type='execution'):
+        largest_instance = None
+        best_node = None
+        if capacity_type != 'execution':
+            for nodes in ('hybrid_nodes', 'execution_nodes'):
+                if instance_group in self.nodes[nodes]:
+                    best_node = self.nodes[nodes][instance_group].best_node()
+            if best_node and best_node.jobs_running == 0:
+                largest_instace = best_node
+        else:
+            # looking for control capacity
+            for nodes in ('hybrid_nodes', 'control_nodes'):
+                if instance_group in self.nodes[nodes]:
+                    best_node = self.nodes[nodes][instance_group].best_node()
+                    if best_node:
+                        break
+            if best_node and best_node.jobs_running == 0:
+                largest_instace = best_node
+        if largest_instance:
+            return largest_instance.hostname
+        return None
+
+    def fit_task_to_most_remaining_capacity_instance(self, task, instance_group, capacity_type, impact=None):
```

Comment:
After tinkering with this a bit today, I have to say that the _old_ `fit_task_to_most_remaining_capacity_instance` method was _beautifully_ isolated and easy to read. It took in 2 arguments - a job, and a list of instances. It was a _static_ method, meaning that it didn't require initializing a class or anything. It got the `capacity_type` from the task itself.

Was this method _in the right place_? No. Attaching it to the manager was confusing. IMO, a better place would be a new module with shared methods, like `awx/main/scheduler/fitting.py`, where it has nothing other than shared static methods.

I'm pro-refactoring, but I'm reading down into this method and seeing new loop structures, and it doesn't seem to be getting any more simple. A list of instances to consider is simple.